tests/load/locustfiles/stored.py

```diff
@@ -0,0 +1,394 @@
+        if not self.host:
+            raise LocustError("'host' value is unavailable.")
+
+        channel_count = random.randint(1, 5)
+
+        ws = self.ws = websocket.WebSocket()
+        self.ws.connect(self.host)
+        self.send_hello(ws)
+        self.on_ws_message(ws, ws.recv())
+        for i in range(channel_count):
+            self.subscribe()
+            self.on_ws_message(ws, ws.recv())
+        self.ws.close()
+
+    def post_notification(self, endpoint_url: str) -> None:
+        """Send a notification to Autopush.
+
+        Args:
+            endpoint_url: A channel destination endpoint url
+        Raises:
+            ZeroStatusRequestError: In the event that Locust experiences a network issue while
+                                    sending a notification.
+        """
+        message_type: str = "notification"
+        # Prefix random message with 'TestData' to more easily differentiate the payload
+        data: str = "TestData" + "".join(
+            [
+                random.choice(string.ascii_letters + string.digits)
+                for i in range(0, random.randrange(1024, 4096, 2) - 8)
+            ]
+        )
+
+        record = NotificationRecord(send_time=time.perf_counter(), data=data)
+        self.notification_records[sha1(data.encode()).digest()] = record
+
+        with self.client.post(
+            url=endpoint_url,
+            name=message_type,
+            data=data,
+            headers=self.REST_HEADERS,
+            catch_response=True,
+        ) as response:
+            if response.status_code == 0:
+                raise ZeroStatusRequestError()
+            if response.status_code != 201:
+                response.failure(f"{response.status_code=}, expected 201, {response.text=}")
+
+    def recv(self, data: str) -> Message | None:
+        """Verify the contents of an Autopush message and report response statistics to Locust.
+
+        Args:
+            data: utf-8 data received from the server
+        """
+        recv_time: float = time.perf_counter()
+        exception: str | None = None
+        message: Message | None = None
+        message_type: str = "unknown"
+        record: Record | None = None
+        response_time: float = 0
+
+        try:
+            message_dict: dict[str, Any] = json.loads(data)
+            message_type = message_dict.get("messageType", "unknown")
+            match message_type:
+                case "hello":
+                    message = HelloMessage(**message_dict)
+                    record = self.hello_record
+                case "notification":
+                    message = NotificationMessage(**message_dict)
+                    message_data: str = message.data
+                    decode_data: str = base64.urlsafe_b64decode(message_data + "===").decode(
+                        "utf8"
+                    )
+                    record = self.notification_records.pop(
```

Comment:
I think, if my memory serves me, in the other user class, the race condition occurs because we are continuously connected. We are processing the recv on a different thread before we finish the send.

But we aren't continuously connected in this scenario right? So the chance of this problem appearing is way lower.

Reply:
Yup :100: (coincidentally I finally remembered this was the exact reason at some point late yesterday)

Reply:
Does it make sense to add a comment to the code so that this knowledge isn't lost? It might make future person's life a bit easier.

Reply:
I think it'd be best to document the behavior in the `locustfile` module's usage where it's not totally obvious the collection's access isn't exclusive -- alternatively that could also use the dict but the access would need to be guarded with a lock.